File: Android/common/base/src/main/java/io/agora/scene/base/component/BaseViewModel.java
package io.agora.scene.base.component;

import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.ViewModel;

public class BaseViewModel extends ViewModel {
    /**
     * 页面回调
     */
    ISingleCallback<Integer, Object> iSingleCallback = null;

    LifecycleOwner lifecycleOwner = null;

    public void setLifecycleOwner(LifecycleOwner lifecycleOwner) {
        this.lifecycleOwner = lifecycleOwner;
    }

    public ISingleCallback<Integer, Object> getISingleCallback() {
        return iSingleCallback;
    }

    public void setISingleCallback(ISingleCallback<Integer, Object> iSingleCallback) {
        this.iSingleCallback = iSingleCallback;
    }
}

File: Android/scenes/joy/src/main/java/io/agora/scene/joy/base/BaseViewModel.kt
package io.agora.scene.joy.base

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

open class BaseViewModel : ViewModel() {
    private val scope = CoroutineScope(Job() + Dispatchers.Main)

    val loadingLiveData = SingleLiveEvent<Boolean>()
    val errorLiveData = SingleLiveEvent<Throwable>()

    fun launch(
        block: suspend () -> Unit,
        error: suspend (Throwable) -> Unit,
        complete: suspend () -> Unit
    ) {
        loadingLiveData.postValue(true)
        scope.launch(Dispatchers.Main) {
            try {
                block()
            } catch (e: Exception) {
                error(e)
            } finally {
                complete()
            }
        }
    }
}